sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala

```diff
@@ -170,4 +174,27 @@ abstract class QueryPlan[PlanType <: TreeNode[PlanType]] extends TreeNode[PlanTy
   protected def statePrefix = if (missingInput.nonEmpty && children.nonEmpty) "!" else ""
 
   override def simpleString: String = statePrefix + super.simpleString
+
+  protected def childrenJson: JValue = {
+    if (children.isEmpty) {
+      JNothing
+    } else if (children.length == 1) {
+      "child" -> children.head.jsonValue
+    } else if (children.length == 2) {
+      ("left" -> children(0).jsonValue) ~ ("right" -> children(1).jsonValue)
+    } else {
+      "children" -> children.map(_.jsonValue)
```

Comment:
I think that storing this as nested JSON is going to make this very hard to query (think about what Spark SQL's JSON schema inference is going to end up with if we use it on this data).  Did you look at @nongli 's proposed JSON?  I think keeping the tree structure implicit is going to give you a schema that is much easier to query.